src/main/scala/org/apache/spark/shuffle/ucx/UcxShuffleTransport.scala

```diff
@@ -0,0 +1,138 @@
+ * it refers to,so when requested this MetadataBlockId, transport could know that soon blockId
+ * would be fetched, and may do some optimizations (pinning, pre-fetching, etc).
+ */
+trait MetadataBlockId extends BlockId {
+  def getBlockId: BlockId
+}
+
+trait Block {
+  // Transport will call this method when it would need an actual block memory.
+  def getMemoryBlock: UcxMemoryBlock
+
+  // Called when the transport is done with the Memory, so we can unmap it for example
+  // the transport is not to use this Memory anymore
+  def doneWithMemory(mem: UcxMemoryBlock)
+}
+
+/**
+ * Async operation callbacks
+ */
+trait OperationCallback {
+  def onSuccess()
+  def onFailure(throwable: Throwable)
+}
+
+/**
+ * Transport flow example:
+ * val transport = new UcxShuffleTransport()
+ * transport.init()
+ *
+ * Mapper/writer:
+ * val cookies = transport.register(dataBlockIds, blocks)
+ * val metadataBlockIds = dataBlockIds.map(dataBlockId => new MetadataBlockId(dataBlockId))
+ * cookies.foreach(cookie => cookie.writeToMemory(metadataBlocks))
+ * transport.register(metadataBlockIds, metadataBlocks) // we don't care for metadata cookies
+ *
+ *
+ * Reducer:
+ * 1. First need to fetch metadata for blockIds:
+ * transport.fetchBlocksByBlockIds(remoteExecutor, metadataBlockIds, resultBuffer, callback)
+ * 2. Deserialize cookies from result buffer:
+ * val cookies = blockIds.map(_ => transport.getCookieFromMemory(resultBuffer))
+ * 3. Fetch blocks by cookies:
+ * transport.fetchBlocksByCookies(remoteExecutor, dataBlockIds, cookies, resultBuffer, callback)
+ * 4. Progress communications:
+ * while(noMoreBlocks) { transport.progress() }
+ *
+ *
+ * transport.unregister(blockIds)
+ * transport.close()
+ */
+trait UcxShuffleTransport {
+
+  /**
+   * Initialize transport resources
+   */
+  def init()
+
+  /**
+   * Close all transport resources
+   */
+  def close()
+
+  /**
+   * Deserialize cookie from memory
+   */
+  def cookieFromMemory(memoryBlock: UcxMemoryBlock): Cookie
+
+  /**
+   * Registers blockId for either metadata blocks, or data blocks on SERVER side.
+   */
+  def register(blockIds: Seq[BlockId], blocks: Seq[Block]): Seq[Cookie]
+
+  /**
+   * Change location of underlying blockId in memory
+   */
+  def mutate(blockIds: BlockId, block: Block, callback: OperationCallback)
+
+  /**
+   * Indicate that this blockId is not needed any more by an application
+   */
+  def unregister(blockIds: Seq[BlockId])
+
+  /**
+   * Fetch remote blocks by blockIds
+   */
+  def fetchBlocksByBlockIds(remoteHost: String, blockIds: Seq[BlockId],
+                            resultBuffer: UcxMemoryBlock, cb: OperationCallback)
+
+  def fetchBlocksByBlockIds(remoteHost: String, blockIds: Seq[BlockId],
+                            resultBuffers: Seq[UcxMemoryBlock], cb: OperationCallback)
+
+  /**
+   * Fetch remote blocks by cookies
+   */
+  def fetchBlocksByCookies(remoteHost: String, blockIds: Seq[BlockId], cookies: Seq[Cookie],
```

Comment:
If we are going to add `blockId` here, should we merge it with `fetchBlocksByBlockId`, and make the cookie optional.

Reply:
I think we need to have 2 methods and always pass cookies, even though the may be internally empty.

Reply:
Can you add descriptions on what each method is supposed to do? we have two variants of each method, and I think they do very different things..  one-to-one vs many-to-one (is that right)? If so, maybe we should call those differently also.

As it stands, `fetchBlocksByCookies` is really `fetchBlocksByBlockIdOrCookies`.  If we are going to have a `fetchBlocksByBlockIds` then `fetchBlocksByCookies` shouldn't take a `BlockId`. I know we've talked about this some, but I am still not sure why a caller can't just just use `fetchBlocksByCookies` if it wants, or `fetchBlocksByBlockIds` if it wants. Both things should work... meaning if you really need the `blockId` in some cases, then in the case where `Cookie` would be nothing, it really should just incorporate the `blockId` so you can make that call within the transport.

Reply:
So, for now, i'll remove second `many-to-many` since there's no optimization in transport for this yet. You can always have `many-to`many` by calling N times `many-to-one`